### ISSUE TITLE
[Question] purpose and encoding of `ioe_filename`

### ISSUE THREAD

**Issue by hasufell** [haskell/core-libraries-committee#189]
[`IOException`](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-IO-Exception.html#v:ioe_filename) has the following definition:

```hs
-- |Exceptions that occur in the @IO@ monad.
-- An @IOException@ records a more specific error type, a descriptive
-- string and maybe the handle that was used when the error was
-- flagged.
data IOException
 = IOError {
     ioe_handle   :: Maybe Handle,   -- the handle used by the action flagging
                                     -- the error.
     ioe_type     :: IOErrorType,    -- what it was.
     ioe_location :: String,         -- location.
     ioe_description :: String,      -- error type specific information.
     ioe_errno    :: Maybe CInt,     -- errno leading to this error, if any.
     ioe_filename :: Maybe FilePath  -- filename the error is related to.
   }
```

Since `ioe_filename` is `Maybe FilePath`, APIs that use `ByteString` or the new `OsPath` have an awkward situation, where they have to decide what to do:

1.  set the field to `Nothing` and be done
2. just use `unpack` (so convert `[Word8]` to `[Char]`)... printing may look like garbage on UTF-8 configured systems
3. use e.g. [decodeFS](https://hackage.haskell.org/package/filepath-1.4.100.3/docs/System-OsPath.html#v:decodeFS), which is the closest match to base API behavior at the FFI boundary, but it's not well-defined for non-utf8 encodings
4. use something like `decodeWith (mkUTF8 TransliterateCodingFailure)`, which never fails and replaces with the closest visual match upon an illegal sequences

This confusion has stirred up debate over at unix: https://github.com/haskell/unix/pull/279

The main question here really is: **should `ioe_filename` allow the caller to recover the original filename or is it just a descriptive field?**

And then: if the former, which encoding do we pick? Right now, the unfortunate situation is that `System.Posix.PosixPath.FilePath` uses method 4 and `System.Posix.ByteString.FilePath` uses method 2. That's clearly undesirable. If we go for method 2 for both, how does the caller know? IMO, this needs to be documented in base. No one will look for this information in the unix modules, nor anticipate that behavior.

@bgamari @mpickering ... do GHC devs have an opinion?

**Comment by bgamari** [comment]
In my opinion `ioe_filename` should merely be a descriptive field to aid in debugging. I think it would be fine to document this fact and would be happy to do so if there is agreement.

**Comment by hasufell** [comment]
> should merely be a descriptive field to aid in debugging

In that case, the better decision in unix would then be to use [`decodeFS`](https://hackage.haskell.org/package/filepath-1.4.100.4/docs/System-OsString-Posix.html#v:decodeFS), which mimics what base does at the outer FFI layer. However, it can fail... I guess we can just catch exceptions and then set `ioe_filename = Nothong` in that case?

**Comment by hasufell** [comment]
Alternative implemented here: https://github.com/haskell/unix/pull/289

**Comment by hasufell** [comment]
> Alternative implemented here: [haskell/unix#289](https://github.com/haskell/unix/pull/289)

We're settling for this solution. 

It may break some users test suites if their filenames contain non-ascii and if they inspect `IOException`.

**Comment by hasufell** [comment]
documentation PR sent for base: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10964
